File: iris/monitors/auditd.py
"""
Auditd monitor — tails /var/log/audit/audit.log, groups records by serial number,
and emits horus events for exec / file-change / network audit keys.

Replaces: fim.py (file changes), process.py (exec), network.py (connections).
Requires auditd rules written by the installer (see install.sh).
Zero inotify watches, zero psutil polling — the kernel audit subsystem does it all.
"""

import logging
import queue
import re
import subprocess
import threading
from pathlib import Path

from iris.config import Config

logger = logging.getLogger(__name__)

AUDIT_LOG = Path("/var/log/audit/audit.log")

_SUSPICIOUS_PORTS = {4444, 5555, 1337, 31337, 6666, 9001}
_BLACKLISTED_CMDS = {"nc", "ncat", "netcat", "socat", "mimikatz", "msfconsole", "msfvenom"}
_SUSPICIOUS_PATHS = ("/tmp/", "/dev/shm/", "/var/tmp/")

_RE_HEADER = re.compile(r"^type=(\S+) msg=audit\(\d+\.\d+:(\d+)\):")
_RE_FIELD   = re.compile(r'(\w+)=(?:"([^"]*?)"|(\S+))')


def _parse(line: str) -> dict:
    m = _RE_HEADER.match(line)
    if not m:
        return {}
    rec = {"type": m.group(1), "_serial": m.group(2)}
    for k, v_q, v_b in _RE_FIELD.findall(line[m.end():]):
        rec[k] = v_q if v_q != "" else v_b
    return rec


class AuditdMonitor:
    def __init__(self, config: Config, event_queue: queue.Queue) -> None:
        self._q = event_queue
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._groups: dict[str, dict] = {}  # serial → {type: rec}

    def start(self) -> None:
        if not AUDIT_LOG.exists():
            logger.warning("AuditdMonitor: %s not found — auditd not installed or not running", AUDIT_LOG)
            return
        self._thread = threading.Thread(target=self._run, name="iris-auditd", daemon=True)
        self._thread.start()
        logger.info("Auditd monitor started (tailing %s)", AUDIT_LOG)

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=5)
        logger.info("Auditd monitor stopped")

    def _run(self) -> None:
        try:
            proc = subprocess.Popen(
                ["tail", "-f", "-n", "0", str(AUDIT_LOG)],
                stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
                text=True, errors="replace",
            )
        except Exception as exc:
            logger.error("AuditdMonitor: failed to tail %s: %s", AUDIT_LOG, exc)
            return

        try:
            while not self._stop.is_set():
                line = proc.stdout.readline()
                if not line:
                    if proc.poll() is not None:
                        break
                    continue
                self._handle_line(line.rstrip())
        finally:
            proc.terminate()
            try:
                proc.wait(timeout=3)
            except subprocess.TimeoutExpired:
                proc.kill()

    def _handle_line(self, line: str) -> None:
        rec = _parse(line)
        if not rec:
            return

        serial = rec["_serial"]
        rtype  = rec["type"]

        if rtype == "EOE":
            group = self._groups.pop(serial, None)
            if group:
                self._process(group)
            return

        self._groups.setdefault(serial, {})[rtype] = rec

        # ponytail: bound memory — drop oldest if too many open groups
        if len(self._groups) > 500:
            for old in sorted(self._groups)[:100]:
                self._groups.pop(old, None)

    def _process(self, group: dict) -> None:
        syscall = group.get("SYSCALL", {})
        key     = syscall.get("key", "")

        if not key.startswith("horus_"):
            return
        if syscall.get("success", "yes") != "yes":
            return

        exe  = syscall.get("exe", "").strip('"')
        comm = syscall.get("comm", "").strip('"')
        uid  = syscall.get("uid", "?")

        if key == "horus_exec":
            execve = group.get("EXECVE", {})
            argc   = int(execve.get("argc", 0) or 0)
            args   = [execve.get(f"a{i}", "") for i in range(min(argc, 8))]
            cmdline = " ".join(args)

            if comm.lower() in _BLACKLISTED_CMDS:
                sev = "high"
            elif any(exe.startswith(p) for p in _SUSPICIOUS_PATHS):
                sev = "medium"
            else:
                return  # routine exec — ignore

            self._emit("suspicious_process", sev,
                f"Suspicious exec: {comm or exe}",
                {"exe": exe, "comm": comm, "cmdline": cmdline[:300], "uid": uid})

        elif key == "horus_fim":
            path_rec = group.get("PATH", {})
            path = path_rec.get("name", "").strip('"')
            if not path:
                return
            sev = "high" if path.startswith(("/etc/", "/root/")) else "medium"
            self._emit("file_change", sev,
                f"File modified: {path}",
                {"path": path, "exe": exe, "uid": uid})

        elif key == "horus_net":
            # Port from hex saddr: AF_INET sockaddr = 02 00 PPPP AAAA...
            saddr = syscall.get("saddr", "")
            port  = None
            try:
                if len(saddr) >= 8 and saddr[:4] in ("0200", "0A00"):
                    port = int(saddr[4:8], 16)
            except ValueError:
                pass

            sev = "high" if port in _SUSPICIOUS_PORTS else "low"
            if sev == "low":
                return  # skip routine connections

            self._emit("network_connection", sev,
                f"Connection to suspicious port {port} from {comm or exe}",
                {"exe": exe, "comm": comm, "uid": uid, "dest_port": port})

    def _emit(self, event_type: str, severity: str, title: str, payload: dict) -> None:
        try:
            self._q.put_nowait({"event_type": event_type, "severity": severity,
                                "title": title, "payload": payload})
        except queue.Full:
            logger.warning("auditd monitor: queue full, dropping event")

File: iris-rs/src/monitors/auditd.rs
//! Auditd monitor; tails /var/log/audit/audit.log, groups records by serial number, and
//! emits events for exec / file-change / network audit keys written by the installer.
//!
//! Replaces the old inotify FIM, procfs process poller, and TCP poller. Zero inotify watches
//! (the cause of the /home OOM), zero polling: the kernel audit subsystem does it all.
//! Requires the auditd rules the installer writes to /etc/audit/rules.d/horus.rules.

use regex::Regex;
use serde_json::{json, Value};
use std::collections::HashMap;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::sync::mpsc;
use tokio::task::JoinHandle;

use crate::config::Config;

const AUDIT_LOG: &str = "/var/log/audit/audit.log";
const MAX_OPEN_GROUPS: usize = 500;

const SUSPICIOUS_PORTS: &[u16] = &[4444, 5555, 1337, 31337, 6666, 9001];
const BLACKLISTED_CMDS: &[&str] = &["nc", "ncat", "netcat", "socat", "mimikatz", "msfconsole", "msfvenom"];
const SUSPICIOUS_PATHS: &[&str] = &["/tmp/", "/dev/shm/", "/var/tmp/"];

pub fn spawn(_config: Config, tx: mpsc::Sender<Value>) -> JoinHandle<()> {
    tokio::spawn(async move {
        if !std::path::Path::new(AUDIT_LOG).exists() {
            tracing::warn!("auditd log {} not found; auditd monitor disabled (install auditd)", AUDIT_LOG);
            return;
        }

        let mut child = match tokio::process::Command::new("tail")
            .args(["-f", "-n", "0", AUDIT_LOG])
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::null())
            .spawn()
        {
            Ok(c) => c,
            Err(e) => {
                tracing::error!("auditd monitor: cannot tail {}: {}", AUDIT_LOG, e);
                return;
            }
        };

        let stdout = match child.stdout.take() {
            Some(s) => s,
            None => return,
        };
        let mut lines = BufReader::new(stdout).lines();
        let parser = Parser::new();
        // serial → (type → fields)
        let mut groups: HashMap<String, HashMap<String, HashMap<String, String>>> = HashMap::new();
        tracing::info!("Auditd monitor started (tailing {})", AUDIT_LOG);

        while let Ok(Some(line)) = lines.next_line().await {
            for evt in parser.handle_line(&line, &mut groups) {
                if tx.send(evt).await.is_err() {
                    let _ = child.kill().await;
                    return;
                }
            }
        }
        let _ = child.kill().await;
    })
}

struct Parser {
    header: Regex,
    field: Regex,
}

impl Parser {
    fn new() -> Self {
        Self {
            header: Regex::new(r"^type=(\S+) msg=audit\(\d+\.\d+:(\d+)\):").unwrap(),
            // key="quoted value" OR key=bareword
            field: Regex::new(r#"(\w+)=(?:"([^"]*)"|(\S+))"#).unwrap(),
        }
    }

    fn parse(&self, line: &str) -> Option<(String, String, HashMap<String, String>)> {
        let h = self.header.captures(line)?;
        let rtype = h[1].to_string();
        let serial = h[2].to_string();
        let rest = &line[h.get(0).unwrap().end()..];
        let mut fields = HashMap::new();
        for c in self.field.captures_iter(rest) {
            let k = c[1].to_string();
            let v = c.get(2).or_else(|| c.get(3)).map(|m| m.as_str()).unwrap_or("");
            fields.insert(k, v.to_string());
        }
        Some((rtype, serial, fields))
    }

    fn handle_line(
        &self,
        line: &str,
        groups: &mut HashMap<String, HashMap<String, HashMap<String, String>>>,
    ) -> Vec<Value> {
        let Some((rtype, serial, fields)) = self.parse(line) else { return vec![] };

        if rtype == "EOE" {
            if let Some(group) = groups.remove(&serial) {
                return process(&group).into_iter().collect();
            }
            return vec![];
        }

        groups.entry(serial).or_default().insert(rtype, fields);

        // ponytail: bound memory; audit records without a clean EOE would leak otherwise.
        if groups.len() > MAX_OPEN_GROUPS {
            let mut keys: Vec<String> = groups.keys().cloned().collect();
            keys.sort();
            for k in keys.into_iter().take(100) {
                groups.remove(&k);
            }
        }
        vec![]
    }
}

fn get<'a>(group: &'a HashMap<String, HashMap<String, String>>, rtype: &str, key: &str) -> &'a str {
    group.get(rtype).and_then(|m| m.get(key)).map(String::as_str).unwrap_or("")
}

fn process(group: &HashMap<String, HashMap<String, String>>) -> Option<Value> {
    let key = get(group, "SYSCALL", "key");
    if !key.starts_with("horus_") {
        return None;
    }
    // auditd records success as "yes"/"no"; default present-but-absent to success.
    if group.get("SYSCALL").and_then(|m| m.get("success")).map(String::as_str).unwrap_or("yes") != "yes" {
        return None;
    }

    let exe = get(group, "SYSCALL", "exe");
    let comm = get(group, "SYSCALL", "comm");
    let uid = get(group, "SYSCALL", "uid");

    match key {
        "horus_exec" => {
            let comm_l = comm.to_lowercase();
            let sev = if BLACKLISTED_CMDS.contains(&comm_l.as_str()) {
                "high"
            } else if SUSPICIOUS_PATHS.iter().any(|p| exe.starts_with(p)) {
                "medium"
            } else {
                return None; // routine exec; ignore
            };
            let argc: usize = get(group, "EXECVE", "argc").parse().unwrap_or(0);
            let args: Vec<String> = (0..argc.min(8))
                .map(|i| get(group, "EXECVE", &format!("a{}", i)).to_string())
                .collect();
            let cmdline: String = args.join(" ").chars().take(300).collect();
            Some(json!({
                "event_type": "new_process",
                "severity": sev,
                "title": format!("Suspicious exec: {}", if comm.is_empty() { exe } else { comm }),
                "payload": {"exe": exe, "comm": comm, "cmdline": cmdline, "uid": uid},
            }))
        }
        "horus_fim" => {
            let path = get(group, "PATH", "name");
            if path.is_empty() {
                return None;
            }
            let sev = if path.starts_with("/etc/") || path.starts_with("/root/") { "high" } else { "medium" };
            Some(json!({
                "event_type": "file_change",
                "severity": sev,
                "title": format!("File modified: {}", path),
                "payload": {"path": path, "exe": exe, "uid": uid},
            }))
        }
        "horus_net" => {
            let port = parse_port(get(group, "SYSCALL", "saddr"))?;
            if !SUSPICIOUS_PORTS.contains(&port) {
                return None; // skip routine connections
            }
            Some(json!({
                "event_type": "new_connection",
                "severity": "high",
                "title": format!("Connection to suspicious port {} from {}", port,
                                 if comm.is_empty() { exe } else { comm }),
                "payload": {"exe": exe, "comm": comm, "uid": uid, "dest_port": port},
            }))
        }
        _ => None,
    }
}

/// Extract the port from an audit hex saddr. AF_INET sockaddr = `0200 PPPP AAAAAAAA …`
/// where PPPP is the big-endian port. (0A00 = AF_INET6, port in the same position.)
fn parse_port(saddr: &str) -> Option<u16> {
    if saddr.len() >= 8 && (saddr.starts_with("0200") || saddr.starts_with("0A00")) {
        u16::from_str_radix(&saddr[4..8], 16).ok()
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_port_from_saddr() {
        // port 4444 = 0x115C
        assert_eq!(parse_port("0200115C7F000001"), Some(0x115C));
        assert_eq!(parse_port(""), None);
    }

    #[test]
    fn flags_exec_from_tmp() {
        let parser = Parser::new();
        let mut groups = HashMap::new();
        parser.handle_line(
            r#"type=SYSCALL msg=audit(1.2:99): exe="/tmp/evil" comm="evil" uid="1000" key="horus_exec" success="yes""#,
            &mut groups,
        );
        let evts = parser.handle_line("type=EOE msg=audit(1.2:99):", &mut groups);
        assert_eq!(evts.len(), 1);
        assert_eq!(evts[0]["event_type"], "new_process");
        assert_eq!(evts[0]["severity"], "medium");
    }

    #[test]
    fn ignores_non_horus_keys() {
        let parser = Parser::new();
        let mut groups = HashMap::new();
        parser.handle_line(
            r#"type=SYSCALL msg=audit(1.2:7): exe="/usr/bin/ls" key="other" success="yes""#,
            &mut groups,
        );
        let evts = parser.handle_line("type=EOE msg=audit(1.2:7):", &mut groups);
        assert!(evts.is_empty());
    }
}
